src/execution/operator/join/physical_index_join.cpp

```diff
@@ -144,22 +150,25 @@ void PhysicalIndexJoin::Output(ExecutionContext &context, DataChunk &input, Data
 }
 
 void PhysicalIndexJoin::GetRHSMatches(ExecutionContext &context, DataChunk &input, OperatorState &state_p) const {
+
 	auto &state = (IndexJoinOperatorState &)state_p;
 	auto &art = (ART &)*index;
-	auto &transaction = Transaction::GetTransaction(context.client);
+
+	// generate the keys for this chunk
+	state.keys.clear();
+	ART::GenerateKeys(state.arena_allocator, state.join_keys, state.keys);
```

Comment:
This is much more efficient already, but we can some memory by re-using the same block of memory for our keys.

The `ArenaAllocator` keeps the blocks that it allocates around. So, when you generate keys on every probe, you accumulate all the keys in the probe in your `ArenaAllocator` for the entire probe. So, when you probe 100M tuples, you will have 100M keys in memory at the end of the probe, which is not necessary.

The same is true for index creation. At the end, all keys will be in memory, even though we process only 1024 keys at a time.

Reply:
This might be achieved by adding a method to the arena allocator that resets but not deallocates the data, making it reusable.